### DISCUSSION TITLE
Does malt support Object coordinate?

### DISCUSSION THREAD

**Discussion by HSDSZ** [bnpr/Malt#159] [Q&A] [answered]
The problem of UV coordinate is the texture is distorted when the shape is changed.
Any way to achieve object coordinate in Malt?

Below is the UV coordinate in malt
![malt](https://user-images.githubusercontent.com/36077492/140082058-55770097-700c-4f7d-aeb5-403bb6bc6c4b.png)

Below is the object coordinate in eevee
![eevee](https://user-images.githubusercontent.com/36077492/140082073-4e77d850-b2b6-483e-957f-f748a81dff80.png)

**Comment by pragma37** [comment]
It's not supported, but I'll look into it.
Shouldn't take too much work.

**Reply by aVersionOfReality** [reply]
I'm very interested in getting these too. I use them a lot already.
Especially the ability to get the coords of other objects, like if you select it from the Object Dropdown in the Texture Coords node.

**Comment by pragma37** [comment, marked as answer]
Ooops! Turns out I completely misunderstood this feature.

For some reason I got it mixed with Generated Coordinates. 🤦‍♂️ 
I actually even implemented the feature, but I was debating whether to publish it since I'm having a hard time figuring under what circumstances it could be useful.

For getting local space coordinates you can just use an inline node with:
```glsl
transform_position(inverse(MODEL), POSITION)
````
MODEL is the object matrix, so the inverse converts world space positions to object space positions.

**Reply by HSDSZ** [reply]
I am not sure I understand your code.
Could you upload a screenshot of the nodes for local space coordinate?

**Reply by pragma37** [reply]
(This is the Development branch, in Release, you'll have to set up the COMMON_PIXEL_SHADER Output in the old way)

![imagen](https://user-images.githubusercontent.com/36610999/142477202-812904e1-6458-49f1-b0c4-f86dd45ed3a8.png)

Or if you want to stick to nodes as much as possible:

![imagen](https://user-images.githubusercontent.com/36610999/142477577-22b42dd3-dc63-4660-8f90-d6bc7ed248fc.png)

The matrix inverse is the only function that is not exposed to nodes.
I plan to rework the way arithmetic works in nodes, so it will probably be available at some point.

**Reply by aVersionOfReality** [reply]
Neat. Is it possible to get the Object Coords (or Matrix i guess) of a different Object, like can be done in Cyles/Eevee?

**Reply by pragma37** [reply]
@aVersionOfReality Using drivers. I guess that's what the Blender node uses internally.
Setting up 16 individual drivers sounds like a PITA, but it's probably not too hard to make an operator that automates it.

Ideally, it should be possible to set any matrix property as an object transform.

**Reply by shuanghenyanshang** [reply]
> Neat. Is it possible to get the Object Coords (or Matrix i guess) of a different Object, like can be done in Cyles/Eevee?
My English is not good, this is machine translation, please forgive me. After looking through other answers, I know that Malt can reference color data of geometric nodes, and try to find out the methods implemented in EV&Cy to invoke coordinate control textures of other objects, but it is a little troublesome. For specific Settings, you can see pictures. (Now 2023.5, blender version 3.5, Malt version 1.0.0Beta)
<img width="1920" alt="Snipaste_2023-05-20_03-53-40" src="https://github.com/bnpr/Malt/assets/80493213/7350be28-a702-4aa4-ab2a-8dc8844e3ad0">
